tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/testlaunch/JUnitLaunchShortcut.xtend

```diff
@@ -0,0 +1,53 @@
+package org.testeditor.tcl.dsl.ui.testlaunch
+
+import javax.inject.Inject
+import org.eclipse.core.resources.IProject
+import org.eclipse.core.resources.IResource
+import org.eclipse.core.runtime.Platform
+import org.eclipse.jface.viewers.ISelection
+import org.eclipse.jface.viewers.IStructuredSelection
+import org.eclipse.jface.viewers.StructuredSelection
+import org.eclipse.ui.IEditorPart
+
+/**
+ * Launch shortcut that is specifically crafted to execute Tests based on TCL files
+ */
+class JUnitLaunchShortcut extends org.eclipse.jdt.junit.launcher.JUnitLaunchShortcut {
+
+	val EXTENSION_POINT_LAUNCHER_ID = "org.testeditor.tcl.dsl.ui.tcl_launcher"
+	val EXTENSION_POINT_CLASS_ATTRIBUTE = "class"
+
+	@Inject
+	extension LaunchShortcutUtil launchUtil
+
+	override void launch(ISelection selection, String mode) {
+		if (selection instanceof IStructuredSelection) {
+			launch(selection.firstElement as IResource, mode)
+		}
+	}
+
+	override void launch(IEditorPart editor, String mode) {
+		launch(editor.editorInput.getAdapter(IResource), mode)
+	}
+
+	private def launch(IResource res, String mode) {
+		if (res.isValidForTestrun) {
+			val je = res.javaElementForResource
+			if (je != null) {
+				val selection = new StructuredSelection(je)
+				val successfulLauncher = extensions.findFirst [ // should be a firstThat
```

Comment:
I don't get the logic of `findFirst` here. As I understand it it would mean that I perform launch for each registered launcher until I come across one that returns `true`. Why do I need that? If I have multiple shouldn't I ask `canLaunch(...)` and then call `launch` for the the one that returns `true` or get back to the user to ask which launcher to use if there are multiple?